src/db/mysql/models.rs

```diff
@@ -90,16 +99,80 @@ impl MysqlDb {
         }
     }
 
-    pub(super) fn create_collection(&self, name: &str) -> Result<i32> {
-        // XXX: handle concurrent attempts at inserts
-        let id = self.conn.transaction(|| {
-            sql_query("INSERT INTO collections (name) VALUES (?)")
-                .bind::<Text, _>(name)
-                .execute(&self.conn)?;
-            collections::table.select(last_insert_id).first(&self.conn)
-        })?;
-        self.coll_cache.put(id, name.to_owned())?;
-        Ok(id)
+    /// APIs for collection-level locking
+    ///
+    /// Explicitly lock the matching row in the user_collections table. Read
+    /// locks do SELECT ... LOCK IN SHARE MODE and write locks do SELECT
+    /// ... FOR UPDATE.
+    ///
+    /// In theory it would be possible to use serializable transactions rather
+    /// than explicit locking, but our ops team have expressed concerns about
+    /// the efficiency of that approach at scale.
+    pub fn lock_for_read(&mut self, user_id: HawkIdentifier, collection: &str) -> Result<()> {
+        let user_id = user_id.legacy_id as u32;
+        let collection_id = self
+            .get_collection_id(collection)
+            .or_else(|e| match e.kind() {
+                // If the collection doesn't exist, we still want to start a
+                // transaction so it will continue to not exist.
+                DbErrorKind::CollectionNotFound => Ok(0),
+                _ => Err(e),
+            })?;
+        // If we already have a read or write lock then it's safe to
+        // use it as-is.
+        if let Some(_) = self.session.coll_locks.get(&(user_id, collection_id)) {
+            return Ok(());
+        }
+
+        // XXX: begin transaction here like python?
+        // Lock the db
+        let modified = user_collections::table
+            .select(user_collections::modified)
+            .filter(user_collections::user_id.eq(user_id as i32))
+            .filter(user_collections::collection_id.eq(collection_id))
+            .lock_in_share_mode()
+            .first(&self.conn)
+            .optional()?;
+        if let Some(modified) = modified {
+            self.session
+                .coll_modified_cache
+                .insert((user_id, collection_id), modified);
+        }
+        // XXX: who's responsible for unlocking
```

Comment:
The Python version implements these as context-managers.  What's the rust equivalent, some sort of guard object in the spirit of `MutexGuard`?

In practice, the python version only ever takes the lock at the start of a request, and releases it at the end.  The actual unlocking is implicit in committing the transaction.  So perhaps there's no need for an explicit unlock, the lock is owned by the session/transaction/whatever and it exists through until that gets committed.

Reply:
Rust equivalents are either a Guard w/ a Drop (but lacking error handling during the dropping) or wrapping everything (via a closure) in another separate call. Diesel encourages running transactions via a closure.

However I don't know if actix_web provides a hook to call into a handler function allowing us to run the transaction in that manner (I doubt it does) -- hence deferring these details until I begin integrating the db w/ handlers

I suspect these lock methods will end up beginning the transaction (like Python) w/ an accompanying actix_web middleware responsible for committing them (or rolling back + rendering an error response)

Reply:
A MutexGuard or passing a closure to a function that executes it.